### ISSUE TITLE
[Enhancement] Support for content providers or cloud

### ISSUE THREAD

**Issue by Mondei1** [IacobIonut01/Gallery#320]
**Is your feature request related to a problem? Please describe.**
I want to fully switch to the cloud with client-side encryption done on my phone and a rented storage box. There are FOSS apps that do that (like Round Sync and DroidFS) but their built-in image/video views are not good and Nextcloud's performance is just bad on Android. So I have no good app to view those images and videos in a comfortable way.

**Describe the solution you'd like**
Both Round Sync, DroidFS, DAVx⁵ and Nextcloud can expose themselves as [content providers](https://developer.android.com/guide/topics/providers/content-provider-basics) which enables other apps to access their storage like internal storage. This does address issue #116 by streamlining the interaction with Nextcloud, as it can be done through the installed Nextcloud app itself. Ideally, this solution requires none to minimal adjustments per app, as long as they adhere to the Android API.

Another nice feature would be to distinct between local and cloud files. For example: The Nextcloud app syncs media to my server and removes the local file after it. The app could now show a cloud icon on the thumbnail and load it though the Nextcloud content provider. So you could parent an album to a content provider (e.g. local DCIM folder to Nextcloud/Camera) and compare files. Like Google Photos does.

**Describe alternatives you've considered**
I have no idea how good the performance would be and how much could be cached (like thumbnails). The alternative would be to implement [`rclone`](https://github.com/rclone/rclone) to enable extensive cloud fetching with client-side encryption, caching and huge protocol support or to simply support native protocols like WebDAV (big ask, I know).

**Additional context**
Content provider shown in the AOSP Files app:

![Screenshot_20240201-171805-final](https://github.com/IacobIonut01/Gallery/assets/22984849/45cb09a9-8acd-457c-97e1-1181976e9834)

Thank you for reading so far. I really hope this gets considered one day.

\- Nicolas

**Comment by IacobIonut01** [comment]
Thanks for the feature request!

I would love to add this feature, I'll add it to the list, but patience is needed as right now I'm working on the next release with Photo Editor and lots of fixes.

But definitely will be done.

**Comment by IacobIonut01** [comment]
And about the complexity, it might add some, but it's mostly on the content parsing method and making it compatible with both Android MediaStore and external content providers

**Comment by Mondei1** [comment]
I experimented a bit with [RSAF](https://github.com/chenxiaolong/RSAF) and got it working. I was able to mount a samba share and overlay a crypt layer on top of it. Now it would be great to be able to access these files with a good gallery app.

1. The external storage has some sort of directory structure, but Gallery can't display and navigate folders, only albums.
    * The first option is to let the user select a folder to "import" using the native file app. After allowing the Gallery app to view a folder from an external storage, we can recursively scan its contents for supported files. (preferred, *maybe even with local index cache?*)
    * The second option is the same as the first, but implementing the file picker part inside the app, but this would require full access to the file system. (I assume)
    * Finally, we could implement a folder structure within the app, but I think that would be overkill for the first iteration.
2. Where should external albums be displayed?
    * Along with the other local albums, but with a new icon to indicate that they are managed externally?
    * Or in the Library tab? (preferred)
3. Where should a button to add new external libraries be?
    * On the album page using a fab icon.
    * In the Library tab (preferred)
4. Should thumbnails persist on the device or are they only in-memory?
5. Limit the loading of some albums to a unmetered wifi connection or leave this to apps like RSAF? Because they will most likely be on a remote server.

There also may be some limitations compared to the local media store. I haven't looked at the codebase yet or tried to build it in Android Studio. I don't want to start with a flawed design. Any more thoughts?

**Comment by cyber-psych** [comment]
@IacobIonut01 duplicate/similar of https://github.com/IacobIonut01/Gallery/issues/116 better to close one to avoid clutter ?

**Comment by Mondei1** [comment]
I would argue that this issue is different. I am referring to support for content providers while #116 is suggesting to support Nextcloud specificly via its API (my guess). This issue can solve #116 by using the content provider exposed by the Nextcloud app, but this approach has some limitations. My issue title is maybe a bit misleading...

tl;dr: This issue describes a more general method to support many cloud providers using a standard interface in Android, while #116 talks about specific support for Nextcloud by using its API. Both can be implemented independently.

**Comment by cyber-psych** [comment]
I guess . I'm against introducing internet requiring functions in a gallery app in the first place . Hope fdroid gets a version without these features if the dev does decide to introduce it .

**Comment by Mondei1** [comment]
The neat thing about this approach is that the app itself does not need internet access and thus not the internet permission. The actual internet part is handled by the content provider app (e.g. Nextcloud, RSAF, etc.), while Gallery uses the local/internal Android API to communicate with the other app. In simple terms: A content provider exposes a filesystem that other apps can navigate.

So it will only be available if you decide to install such a provider and connect the Gallery app to it. And as far as I know, no other app currently supports this despite being really useful.

**Comment by AlphaCraft9658** [comment]
Any updates?

**Comment by IacobIonut01** [comment]
> Any updates?

I'm not able to increase the priority yet, however I will get back with updates asap I can take it as a priority

**Comment by olepl** [comment]
Is this part of the latest release? I'm trying to access the photos from my Proton Drive (regular File Manager sees it) but I can't find it anywhere in the app

**Comment by olepl** [comment]
I'll ping @IacobIonut01 because as described in my comment this issue does not seem to be solved for me, perhaps I should open a new issue?
